Agilefreaks/FreaksCatalog-API
# [143] Add filter support to the freaks query node

## Description

Add filters to the freaks query node. You can filter freaks by projects or technologies after next criteria:
`allOf`: freak needs to work on All the the projects / have all the skills
`anyOf`: freak needs to work on at least 1 project from the list / have at least 1 skill from the list

## Related

[[143] Add filter support to the "freaks" query node](https://trello.com/c/gKNOKo3S/143-add-filter-support-to-the-freaks-query-node)

## Steps to Test or Reproduce

Go to Insomnia and execute this graphql query: 
```GraphQl
query Freaks($filter: FreakFilter, $last: Int, $before: String) {
  freaks(filter: $filter, last: $last, before: $before) {
    edges {
      node {
        id
        firstName
        projects {
          id
          name
        }
        technologies 
        {
          id
          name
        }
      }
    }
  }
}
```
You should get a list with all freaks, because no filters are selected.
To apply filters,  go to bottom left at **Query variables** and introduce next JSON code to filter:
```JSON
{
  "filter": {
    "projectIds":{
        "allOf": [1, 2]
     }
  }
}
```
You should get a list with freaks that are working on both projects with id 1 and 2.


```JSON
{
  "filter": {
    "projectIds":{
        "anyOf": [1, 2]
     }
  }
}
```
You should get a list with freaks that are working on any projects with id 1 or 2.


```JSON
{
  "filter": {
    "projectIds":{
      	"allOf": [1],
        "anyOf": [1, 2]     
    }
  }
}
````
You should get a message of error, because these 2 filters are mutually exclusive.


You can filter a freak after technologies by the same way, just replace `projectIds` with `technologyIds`.


Also, you can combine filters like this:
```JSON
{
  "filter": {
    "projectIds":{
      	"allOf": [2,3]
    },
    "technologyIds": {
      "anyOf": [1,2]
    }
  }
}
```
You should get a list with all freaks that are working on projects with ids 2 and 3 and have at least one technology with id 1 or 2.
